### ISSUE TITLE
shuffle index

### ISSUE THREAD

**Issue by Gorazor** [ultralytics/yolov5#2582]
## 🐛 Bug
A  bug exsits in func create_dataloader in utils/dataset.py.  If training not in DDP mode, no shuffle operation in dataloader. if you training data configed as list: list: [path1/images/, path2/images/] and each path's images are different classes or sources. This could be a problem.  (this may not be a big problem if you use mosaic since for mosaic augment other image pathes are random selected)
```
  ......
  sampler = torch.utils.data.distributed.DistributedSampler(dataset) if rank != -1 else None
  loader = torch.utils.data.DataLoader if image_weights else InfiniteDataLoader
  dataloader = loader(dataset,
                      batch_size=batch_size,
                      num_workers=nw,
                      sampler=sampler,
                      pin_memory=True,
                      collate_fn=LoadImagesAndLabels.collate_fn4 if quad else LoadImagesAndLabels.collate_fn)

 ......
```

##  solution
add an argument ''is_train' for shuffle
```
def create_dataloader(path, imgsz, batch_size, stride, opt, hyp=None, augment=False, cache=False, pad=0.0, rect=False,
                      rank=-1, world_size=1, workers=8, image_weights=False, quad=False, prefix='', **is_train=True**):
    ......
    sampler = torch.utils.data.distributed.DistributedSampler(dataset) if rank != -1 else None
    loader = torch.utils.data.DataLoader if image_weights else InfiniteDataLoader
    shuffle = True if is_train and sampler is None else False
    dataloader = loader(dataset,
                        batch_size=batch_size,
                        num_workers=nw,
                        shuffle=shuffle,
                        sampler=sampler,
                        pin_memory=True,
                        collate_fn=LoadImagesAndLabels.collate_fn4 if quad else LoadImagesAndLabels.collate_fn)
   ......
```

**Comment by glenn-jocher** [comment]
@Gorazor yes this is a good point, which has also been brought up before. Though as you mention if you use mosaic=1.0 in your hyperparameters then 75% of the images will still be shuffled.

@Gorazor have you tried to train some baseline datasets against the default code with your proposed change? A quick test might be to finetune VOC with YOLOv5s using the existing code and your proposed code and compare the result. In Colab this usually only takes a few hours, i.e.:
https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb#scrollTo=BSgFCAcMbk1R&line=1&uniqifier=1

```python
# VOC
for b, m in zip([64], ['yolov5s']):  # zip(batch_size, model)
  !python train.py --batch {b} --weights {m}.pt --data voc.yaml --epochs 50 --cache --img 512 --nosave --hyp hyp.finetune.yaml --project VOC --name {m}
```

**Comment by glenn-jocher** [comment]
@Gorazor good news 😃! Your original issue may now be fixed ✅ in PR #5623 by @werner-duvaud. This PR turns on shuffling in the YOLOv5 training DataLoader by default, which was missing until now. This works for all training formats: CPU, Single-GPU, Multi-GPU DDP.

```python
train_loader, dataset = create_dataloader(train_path, imgsz, batch_size // WORLD_SIZE, gs, single_cls,
                                          hyp=hyp, augment=True, cache=opt.cache, rect=opt.rect, rank=LOCAL_RANK,
                                          workers=workers, image_weights=opt.image_weights, quad=opt.quad,
                                          prefix=colorstr('train: '), shuffle=True)  # <--- NEW
```

I evaluated this PR against master on VOC finetuning for 50 epochs, and the results show a slight improvement in most metrics and losses, particularly in objectness loss and mAP@0.5, perhaps indicating that the shuffle addition may help delay overtraining.

<img width="979" alt="Screenshot 2021-11-13 at 13 03 26" src="https://user-images.githubusercontent.com/26833433/141643075-70b06810-4727-4f5c-b03a-f68c34fe6934.png">

https://wandb.ai/glenn-jocher/VOC

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!
